### ISSUE TITLE
Use broadcast instead of flooding the internal network

### ISSUE THREAD

**Issue by jonas-w** [localsend/protocol#1]
In my opinion it would be way better to send broadcasts with an identifier and listen for broadcasts instead of scanning the whole network. Maybe scanning all devices is a "viable" solution in little /24 networks, but i've seen networks as big as a /16 subnet and you definitely don't want to send an http request to 65,534 devices. And even if only 2 people use localsend in that network there nearly would be a whopping 130k requests just for sending a single file to another device.

This would work like so:
Device#1 sends a broadcast with it's alias, http server's portnumber and says hey i'm using localsend.
Device#2 listens for broadcasts and sees Device#1's broadcast message, and connects to the http port defined in the broadcast message and does whatever it needs to do.

I think kdeconnect does it this way, you could just capture the traffic when pairing with kdeconnect to understand it a bit more.

**Comment by jonas-w** [comment]
I've found the source:
https://invent.kde.org/network/kdeconnect-kde/-/blob/master/core/backends/lan/lanlinkprovider.cpp#L121

It basically gets the brodcast address(es) for this network, and generates an identifying packet which tells others your name/id etc. And then sends this UDP packet to the broadcast address(es) of this network. https://en.wikipedia.org/wiki/Multicast_address

But I like the idea of using https for file transfer, as this is currently one of the best ways to send files locally. And it was really really fast when i tried it yesterday.

**Comment by Jipok** [comment]
I had a problem with the fact that the laptop(`#2`(Nice idea with `#Num)`) did not see the phone(`#206`). I decided to look at the program logs in the hope of finding something. And found chilling horror. Because of which I immediately ran to create an issue here, fortunately I'm not the only one who noticed this.
It's funny that I was just thinking how it will affect the future of the IT industry that not everyone studies such fundamental things as networks. (I'm not trying to offend you.) And given the amount of information required by modern programmers, it may not be worth asking everyone.

I suggest you take a free multicast address from this list:
https://www.iana.org/assignments/multicast-addresses/multicast-addresses.xhtml#multicast-addresses-3
Something like 224.0.245.246, which seems to be unused. And of course on port 53317.

For the sake of interest, I wrote a code that implements the proposed mechanism. Although your code is not in python, it may help you:
```python
#!/usr/bin/env python3
import socket
import struct
import select
import time

DISCOVERY_COOLDOWN = 5 # Secunds
MCAST_GRP = '224.0.245.246'
MCAST_PORT = 53317
DESC = """{
  "alias": "%s",
  "deviceModel": "Some",
  "deviceType": "Server"
}""" % (socket.gethostname())

multicast = socket.socket(socket.AF_INET, socket.SOCK_DGRAM, socket.IPPROTO_UDP)
multicast.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
multicast.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_TTL, 2)
multicast.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_LOOP, 1)
multicast.bind((MCAST_GRP, MCAST_PORT))
mreq = struct.pack("4sl", socket.inet_aton(MCAST_GRP), socket.INADDR_ANY)
multicast.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, mreq)

server = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
server.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
server.bind(("0.0.0.0", MCAST_PORT))
server.listen()

last_discovery = 0

while True:
  if time.monotonic() - last_discovery > DISCOVERY_COOLDOWN:
    last_discovery = time.monotonic()
    print(chr(27) + "[2J") # Clear screen
    multicast.sendto(DESC.encode("UTF-8"), (MCAST_GRP, MCAST_PORT))
  rlist = select.select([multicast, server], [], [], 5)[0]
  for sock in rlist:
    if sock == multicast:
      (bytes, address) = multicast.recvfrom(1024)
      data = bytes.decode("UTF-8")
      print("Discovery requested by: %s" % address[0])
      print(data)
      tmp_sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
      tmp_sock.connect(address)
      tmp_sock.send(DESC.encode("UTF-8"))
      tmp_sock.close()
    elif sock == server:
      connection, address = server.accept()
      print("Get response from: %s" % address[0])
      print(connection.recv(4096).decode("UTF-8"))
      connection.close()
```

**Comment by Tienisto** [comment]
Thank you! I will look into it!

**Comment by Tienisto** [comment]
Implemented in `LocalSend v1.7.0`

**Comment by Jipok** [comment]
Nice to see this change. I have never seen dart code. But as I understand it, you are responding to the request with broadcasting, which I do not recommend. UDP packets may be lost. And they are really lost, especially over wifi networks. In the example above, I also initially used only UDP, until I noticed that sometimes clients do not see each other(and I had only 3 devices, one was connected to the router via cable). After that, I began to use the TCP for the answer.
I also had the package sent to the requester only. In the case of broadcasting, the chance of losses is actually higher. Just imagine that 5 WIFI connected clients receive UDP and respond at the same time. That's 25 UDP packets at once, I'm sure some will get lost each time.
I haven't looked at all code, but you already have an http server that works with different endpoints. Just add another one that will accept a POST with a description of the peer.

**Comment by Tienisto** [comment]
Broadcasting continuously every few seconds is actually planned but I was too lazy to implement this now.

Currently, the "announcement" of a new device is triggered multiple (currently 4x) times so it somehow works. Yes, respnding with TCP is a more solid solution, I overlooked this in your example.

**Comment by marcossmtp** [comment]
Is possible to implement that computers in the same router but conected with eth not wifi use the same protocol.
I installed the app in computer in the same network and cant send files nor writting the correct ip.

**Comment by marcossmtp** [comment]
> @marcossmtp
> 
> > Is possible to implement that computers in the same network but conected with eth not wifi use the same protocol?. I installed the app in computer in the same network and cant send files nor writting the correct ip.
> 
> Ethernet or wifi does not matter as long as both devices are in the same local network. Your issue is most likely a firewall issue. See [#149](https://github.com/localsend/localsend/issues/149), [#27](https://github.com/localsend/localsend/issues/27), [#53](https://github.com/localsend/localsend/issues/53)

thanks very much
 and is possible to send files to the the group with multiast into the app?

**Comment by Tienisto** [comment]
@Jipok thanks, I updated the protocol and implemented the TCP response in `v1.8.0`.

@marcossmtp please open a separate issue, I also didn't understand your question. If you mean, that you want to send to multiple devices in parallel, then this is possible since `v1.8.0`.

**Comment by marcossmtp** [comment]
> 

ok, yes that's was my question, thanks.
Going to check 1.8.0
